### PR TITLE
NEXT-29869 Add versioning for Capture, Refund and RefundPosition entities, fixes #3195

### LINKED ISSUE #3195: OrderTransactionCapture entities related to an OrderTransaction are deleted just after completion of Order edit in Admin Panel
### PHP Version

8.1

### Shopware Version

6.5.2.1 (affects all versions >=6.4.12.0)

### Expected behaviour

When an Order is available in the system, and an `OrderTransactionCapture` entity is created as a relation to `OrderTransaction` for this particular Order, then editing this Order via Admin Panel shouldn't affect the `OrderTransactionCapture` entity - it should preserve all of the attribute values and still be available in the system with the same relations prior to Order edit.

### Actual behaviour

When an admin edits an Order, which has `OrderTransactionCapture` related to `OrderTransaction` for the edited order, then once an edition is saved (i.e. the system is merging versions of new order edit and old order) all of the related `OrderTransactionCapture` entities in the database are being deleted.

This is happening, because both `Order` and `OrderTransaction` are versioned entities, while `OrderTransactionCapture` is not, and `OrderTransactionDefinition` do have the `order_transaction.captures` relation using the `CascadeDelete` flag. During the order edit, a second version of `Order` and `OrderTransaction` is created, and once the `merge` process occurs, one version is deleted making non-versioned (while related) `OrderTransactionCapture` be deleted because of the cascade relation definition.

### How to reproduce

1. Create an Order with OrderTransaction
2. Create an `OrderTransactionCapture` (currently, only via API, POST: `/api/order-transaction-capture`) linked to `OrderTransaction` from pt.1 above
3. Confirm the entity entry is visible in the database (`order_transaction_capture` table)
4. Edit and save the order created in pt.1 to trigger the version created + merge process (can be by adding a tracking code)
5. Check for the entity entry (as in pt.3) - there is no entity in the database anymore

The fixed solution would be one of the below:
a) remove the `CascadeDelete` flag on the `order_transaction.captures` relation
b) introduce versioning for `OrderTransactionCapture`, `OrderTransactionCaptureRefund`, and `OrderTransactionCaptureRefundPosition` accordingly fixed

### PR DESCRIPTION
**Pull request by matkowalski**
### 1. Why is this change necessary?

Without the change, the issue reported in #3195 occurs. In short, when the system processes the merge on a versioned `OrderTransaction` entity, then it affects related `OrderTransactionCapture` entities and deletes all of them if there are any linked to `OrderTransaction` (because `OrderTransactionCapture` is not versioned).

If there are existing `OrderTransactionCapture` entities related to an `OrderTransaction`, and when admin edits an`Order` with such a transaction via Administration Panel, then all linked `OrderTransactionCapture` entities are deleted on cascade when an `Order` is saved (i.e. when versions are merged). This results in transaction data loss in the database.

### 2. What does this change do, exactly?

The change adds versioning for `OrderTransactionCapture`, `OrderTransactionCaptureRefund`, and `OrderTransactionCaptureRefundPosition` entities, so that when editing an order a new version for these entities is created. This allows to change data on revised order and the system can relate to a specific version for these entities via `versionId`.

### 3. Describe each step to reproduce the issue or behaviour.

1. Create an Order with OrderTransaction
2. Create an `OrderTransactionCapture` (currently, only via API, POST: `/api/order-transaction-capture`) linked to `OrderTransaction` from pt.1 above
3. Confirm the entity entry is visible in the database (`order_transaction_capture` table)
4. Edit and save the order created in pt.1 to trigger the version created + merge process (can be by adding a tracking code)
5. Check for the entity entry (as in pt.3) - there is no entity in the database anymore

### 4. Please link to the relevant issues (if any).

Relates and fixes #3195

### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [x] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [x] I have written or adjusted the documentation according to my changes
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### <samp>🤖 Generated by Copilot at 5adaed3</samp>

This pull request adds versioning support to the entities related to order transaction captures and refunds. This allows tracking the history and changes of the captures and refunds over time. It modifies the entity definitions, classes, API and database tables to include the version_id fields and references. It also updates the changelog and adds a migration class to apply the changes to the existing data.

### <samp>🤖 Generated by Copilot at 5adaed3</samp>

*  Add a changelog entry for the pull request ([link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-1626b0aa04e331241905751433e93758773cd19bfd5fbec13666f256baf6b484R1-R13))
*  Add version_id fields to the OrderTransactionCapture, OrderTransactionCaptureRefund and OrderTransactionCaptureRefundPosition entities, which are the primary keys of the entities and can be exposed to the API ([link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-f4619adc95f1f056d1680ffaed2e5bd8f15d747dd5c4759d4dd2f5444ba107b2R64), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-b91a97d1c16094354394234c6413ec37aa929147fd4c0164c46cf11edc25e132L62-R67), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-d996c859e01c71cb9b027cd66a8f4ee751dbb21559a86d8b95bf172e5e7c21a1L56-R59))
*  Add reference version fields to the OrderTransactionCaptureRefund and OrderTransactionCaptureRefundPosition entities, which link them to the version_id of the related OrderTransactionCapture and OrderTransactionCaptureRefund entities, respectively, and can be exposed to the API ([link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-b91a97d1c16094354394234c6413ec37aa929147fd4c0164c46cf11edc25e132L62-R67), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-d996c859e01c71cb9b027cd66a8f4ee751dbb21559a86d8b95bf172e5e7c21a1L56-R59))
*  Add properties and methods for the version_id and reference version fields to the corresponding entity classes, which allow accessing and modifying the values of the fields ([link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-94c4f164d1ecac8d6ef35575f2911271999d029d3f7d3f275d87955be9d33067R22-R23), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-94c4f164d1ecac8d6ef35575f2911271999d029d3f7d3f275d87955be9d33067R105-R114), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-749d558c059c3c3a97d5adc3353dbeed5cdbb618fb334485468d8b7908ae5defR22-R23), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-749d558c059c3c3a97d5adc3353dbeed5cdbb618fb334485468d8b7908ae5defR117-R126), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-c70d5343e365e052689fdc39cd437f6163d5897d525f7e68589cebaa6e366ed2R21-R22), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-c70d5343e365e052689fdc39cd437f6163d5897d525f7e68589cebaa6e366ed2R116-R125))
*  Add use statements for the VersionField, ReferenceVersionField and OrderTransactionDefinition classes to the entity definition classes, which are needed to define the version_id and reference version fields ([link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-f4619adc95f1f056d1680ffaed2e5bd8f15d747dd5c4759d4dd2f5444ba107b2R21), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-b91a97d1c16094354394234c6413ec37aa929147fd4c0164c46cf11edc25e132R5), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-b91a97d1c16094354394234c6413ec37aa929147fd4c0164c46cf11edc25e132L18-R22), [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-d996c859e01c71cb9b027cd66a8f4ee751dbb21559a86d8b95bf172e5e7c21a1R19))
*  Add a migration class to the Core namespace, which alters the database tables for the entities to support versioning, changes the primary and foreign keys to rely on the version_id fields, and sets the default version_id values to the live version ID for the existing records (`src/Core/Migration/V6_5/Migration1689856589AddVersioningForOrderTransactionCaptures.php`, [link](https://github.com/shopware/platform/pull/3228/files?diff=unified&w=0#diff-421111a3710d31d7d48d314aff180a86f0a3ad8c543abbef3c62b5c9b1b56f68R1-R94))

### CODE REVIEW

**Comment by silverDuy** [comment]
Hello @matkowalski,

I want to express my gratitude for your efforts on this PR. Overall, it looks good to me. However, I have a few suggestions and a checklist for you to consider:

1. Before adding the column, checking if it already exists is best to avoid any conflicts.
2. Drop the current PRIMARY KEY and add it again like ADD PRIMARY KEY (`id`, `version_id`)
3. For better data integrity, please add CONSTRAINT and REFERENCES to the parent table's version_id.
4. It would be great if you could write a unit test for the migration file. This will ensure that running the migration multiple times will avoid duplicate data.

Thank you again for your hard work, and I look forward to seeing these improvements in the PR.

**Comment by matkowalski** [comment]
Thanks for the feedback, @silverDuy 

1. Added the check before adding a column
2. Dropping and re-adding primary keys was already covered in the code, but refactored the code a bit to give more clarity
3. Constrains were also added previously but refactored the code into more self-description and granular sequences for clarity
4. Added a test, while this is just a test verifying that the migration can be run multiple times (as you indicated).

Please let me know if this might do. Thanks!

File: src/Core/Checkout/Order/Aggregate/OrderTransactionCapture/OrderTransactionCaptureEntity.php [hunk 1]
```diff
@@ -100,4 +102,14 @@ public function setRefunds(OrderTransactionCaptureRefundCollection $refunds): vo
     {
         $this->refunds = $refunds;
     }
+
+    public function getOrderTransactionVersionId(): string
+    {
+        return $this->orderTransactionVersionId;
+    }
+
+    public function setOrderVersionId(string $orderTransactionVersionId): void
```
Comment (silverDuy) [inline src/Core/Checkout/Order/Aggregate/OrderTransactionCapture/OrderTransactionCaptureEntity.php]:
```suggestion
    public function setOrderTransactionVersionId(string $orderTransactionVersionId): void
```

File: changelog/_unreleased/2023-07-20-add-versioning-to-capture-capturerefund-and-capturerefundposition-entities.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+author: Mateusz Kowalski
+author_email: matt@websolutionsnyc.com
+author_github: @matkowalski
+---
+# Core
+* Added `version_id` field to `OrderTransactionCapture`, `OrderTransactionCaptureRefund` and `OrderTransactionCaptureRefundPosition` entities
+* Changed relations in `OrderTransactionCaptureRefund` and `OrderTransactionCaptureRefundPosition` to support versioned field
+* Added new version related columns to `order_transaction_capture`, `order_transaction_capture_refund` and `order_transaction_capture_refund_position` tables
+* Changed Primary Key for `order_transaction_capture`, `order_transaction_capture_refund` and `order_transaction_capture_refund_position` tables to rely on `id` and `version_id` fields
+* Changed Foreign Key for `order_transaction_capture`, `order_transaction_capture_refund` and `order_transaction_capture_refund_position` tables to support versioned primary keys
```
Comment (silverDuy) [inline changelog/_unreleased/2023-07-20-add-versioning-to-capture-capturerefund-and-capturerefundposition-entities.md]: Some suggestions for the changelog:

```md
* Added `VersionField` into `OrderTransactionCaptureDefinition`, `OrderTransactionCaptureRefundDefinition` and `OrderTransactionCaptureRefundPositionDefinition`
* Added property `orderTransactionVersionId` into `OrderTransactionCaptureEntity`
* Added `ReferenceVersionField` to reference to `OrderTransactionDefinition` into `OrderTransactionCaptureRefundDefinition`
* Added property `captureVersionId` into `OrderTransactionCaptureRefundEntity`
* Added `ReferenceVersionField` to reference to `OrderTransactionCaptureRefundDefinition` into `OrderTransactionCaptureRefundPositionDefinition`
* Added property `refundVersionId` into `OrderTransactionCaptureRefundPositionEntity`
* Added migration `Migration1689856589AddVersioningForOrderTransactionCaptures` to:
    * Added new version related columns to `order_transaction_capture`, `order_transaction_capture_refund` and `order_transaction_capture_refund_position` tables
    * Changed Primary Key for `order_transaction_capture`, `order_transaction_capture_refund` and `order_transaction_capture_refund_position` tables to rely on `id` and `version_id` fields
    * Changed Foreign Key for `order_transaction_capture`, `order_transaction_capture_refund` and `order_transaction_capture_refund_position` tables to support versioned primary keys
```

**Comment by matkowalski** [comment]
@silverDuy is there anything else I can do to move the processing of the current Pull Request? The reported bug and issue in #3195 prevent us from delivery of Payment Method solutions at Shopware Store for US Market, which does rely on Capture and Refund entities for their features.

**Comment by silverDuy** [comment]
@matkowalski Hi, I apologize for the delay in responding. I've made some minor suggestions, and everything looks good to me 👍.

Please review my suggestions, and if they make sense to you, kindly address them. I'll take care of the remaining afterwards.

**Comment by matkowalski** [comment]
@silverDuy all done, thanks for the feedback and actions, looking forward to the progress.

**Comment by silverDuy** [comment]
@matkowalski Thank you. We will do the rest.

File: src/Core/Checkout/Order/Aggregate/OrderTransactionCaptureRefund/OrderTransactionCaptureRefundDefinition.php [hunk 3]
```diff
@@ -59,7 +62,9 @@ protected function defineFields(): FieldCollection
 
         return new FieldCollection([
             (new IdField('id', 'id'))->addFlags(new ApiAware(), new PrimaryKey(), new Required()),
+            (new VersionField())->addFlags(new ApiAware()),
             (new FkField('capture_id', 'captureId', OrderTransactionCaptureDefinition::class))->addFlags(new ApiAware(), new Required()),
+            (new ReferenceVersionField(OrderTransactionDefinition::class, 'capture_version_id'))->addFlags(new ApiAware(), new Required()),
```
Comment (nguyenytran) [inline src/Core/Checkout/Order/Aggregate/OrderTransactionCaptureRefund/OrderTransactionCaptureRefundDefinition.php]:
```suggestion
            (new ReferenceVersionField(OrderTransactionCaptureDefinition::class, 'capture_version_id'))->addFlags(new ApiAware(), new Required()),
```

please run `bin/console dal:validate` to check DAL

Comment (matkowalski) [reply src/Core/Checkout/Order/Aggregate/OrderTransactionCaptureRefund/OrderTransactionCaptureRefundDefinition.php]: thanks! fixed, `dal:validate` does not report anything else, so this should be all good now

**Comment by keulinho** [comment]
Thanks for your contribution, your PR was merged! :tada: :blue_heart:

**Pull request closed without merging**